r/R/dotplot.R

```diff
@@ -1,35 +1,84 @@
+#' Title
+#'
+#' @param req {body: {
+#'               useMarkerGenes: True/False determines whether to use marker genes or predefined genes
+#'               numberOfMarkers: Int. Number of marker genes to use
+#'               customGenesList: List of Strings. List of marker genes to use
+#'               groupBy: Cellsets to show in the plot. Determines whether to show Louvain/Samples/Custom
+#'               filterBy: Cellsets to subset the experiment with.
+#'               isFilterByAll: Bool value. If true, the experiment will be subsetted to all the cellSets ids.
+#'              }
+#'            }
+#' @param data
+#'
+#' @return
+#' @export
+#'
+#' @examples
 runDotPlot <- function(req, data) {
-  type <- req$body$type
-  req_genes <- req$body$genes
+  useMarkerGenes <- req$body$useMarkerGenes
+  data$dotplot.groups <- NA
+  group_by_cell_sets <- req$body$groupBy$children
+  filter_by <- req$body$filterBy
+  apply_filter <- req$body$applyFilter
 
-  # construct clusters from cell sets
-  data$custom <- NA
-  cells_id <- data$cells_id
-  cell_sets <- req$body$cellSets$children
+  if (length(group_by_cell_sets) < 1) {
+    message("The requested Cell Sets are empty. Returning empty results.")
+    return(list())
+  }
 
-  for (i in seq_along(cell_sets)) {
-    set <- cell_sets[[i]]
-    filtered_cells <- intersect(set$cellIds, cells_id)
-    data$custom[cells_id %in% filtered_cells] <- i
+  #Collect ids to subset object
+  if (apply_filter) {
+    subsetIds <- filter_by$cellIds
+  } else {
+    subsetIds <- list()
+    for (child in filter_by$children) {
+      subsetIds <- append(subsetIds, child$cellIds)
+    }
```

Comment:
```suggestion
    subset_ids <- unlist(lapply(filter_by$children, `[[`, 'cellIds'))
```

untested, could break into two lines if you use. Not sure why your creating a list of lists of cell ids instead of a vector?

Reply:
It's not a list of lists, the append function will add the elements of each cellIds list at the end of the subsetIds, returning a one level list of cells Id. That being said, I like your solution more :D